nvdaHelper/remote/COMProxyRegistration.cpp

```diff
@@ -0,0 +1,167 @@
+	}
+	// Create and activate an activation context using the manifest in the proxy dll 
+	// to temporarily register the proxy dll's class object
+	ACTCTX actCtx={0};
+	actCtx.cbSize=sizeof(actCtx);
+	actCtx.dwFlags=ACTCTX_FLAG_HMODULE_VALID|ACTCTX_FLAG_RESOURCE_NAME_VALID;
+	actCtx.lpResourceName=MAKEINTRESOURCE(2);
+	actCtx.hModule=dllHandle;
+	HANDLE hActCtx=CreateActCtx(&actCtx);
+	if(hActCtx==NULL) {
+		LOG_ERROR(L"Could not create activation context for "<<dllPath);
+		FreeLibrary(dllHandle);
+		return nullptr;
+	}
+	ULONG_PTR actCtxCookie;
+	if(!ActivateActCtx(hActCtx,&actCtxCookie)) {
+		LOG_ERROR(L"Error activating activation context for "<<dllPath);
+		ReleaseActCtx(hActCtx);
+		FreeLibrary(dllHandle);
+		return nullptr;
+	}
+	// Fetch the class object (which will come from the proxy dll)
+	IUnknown* ClassObjPunk=NULL;
+	res=CoGetClassObject(*pProxyClsid,CLSCTX_INPROC_SERVER,nullptr,IID_IUnknown,(void**)&ClassObjPunk);
+	// From here we no longer need the activation context
+	DeactivateActCtx(0,actCtxCookie);
+	ReleaseActCtx(hActCtx);
+	if(res!=S_OK) {
+		LOG_ERROR(L"Error fetching class object for "<<dllPath<<L", code "<<res);
+		FreeLibrary(dllHandle);
+		return nullptr;
+	}
+	// Re-register the class object with COM now that the activation context is gone.
+	// Keeping the class object available to COM, with COM also handling the life time of the proxy dll now
+	DWORD dwCookie;
+	res=CoRegisterClassObject(regClsid,ClassObjPunk,CLSCTX_INPROC_SERVER,REGCLS_MULTIPLEUSE,&dwCookie);
+	ClassObjPunk->lpVtbl->Release(ClassObjPunk);
+	if(res!=S_OK) {
+		LOG_ERROR(L"Error registering class object for "<<dllPath<<L", code "<<res);
+		FreeLibrary(dllHandle);
+		return nullptr;
+	}
+	COMProxyRegistration_t* reg= new COMProxyRegistration_t();
+	reg->dllPath=dllPath;
+	reg->classObjectRegistrationCookie=dwCookie;
+	// For all interfaces the proxy dll supports, register its CLSID as their proxy stub CLSID
+	ProxyFileInfo** tempInfoPtr=pProxyInfo;
+	while(*tempInfoPtr) {
+		ProxyFileInfo& fileInfo=**tempInfoPtr;
+		for(unsigned short idx=0;idx<fileInfo.TableSize;++idx) {
+			IID iid=*(fileInfo.pStubVtblList[idx]->header.piid);
+			CLSID clsidBackup={0};
+			wstring_convert<codecvt_utf8_utf16<wchar_t>> converter;
+			wstring name=converter.from_bytes(fileInfo.pNamesArray[idx]);
+			// Fetch the old CLSID for this interface if one is set, so we can replace it on deregistration.
+			// If not set, then we'll use the standard marshaler clsid on deregistration.
+			if((res=CoGetPSClsid(iid,&clsidBackup))!=S_OK) {
+				clsidBackup=clsid_standardMarshaler;
+			} else {
+				LOG_DEBUG(L"Backed up existing clsid for interface "<<name);
+			}
+			if((res=CoRegisterPSClsid(iid,regClsid))!=S_OK) {
+				LOG_ERROR(L"Unable to register interface "<<name<<L" with proxy stub "<<dllPath<<L", code "<<res);
+				continue;
+			}
+		reg->psClsidBackups.push_back({name,iid,clsidBackup});
+		LOG_DEBUG(L"Registered interface "<<name);
+		}
+		++tempInfoPtr;

```

Comment:
Is there some docs you can point to on ProxyFileInfo? I'm a little nervous about this code. So `pProxyInfo` points to the start of a memory block containing `ProxyFileInfo*`. This memory block is null terminated. I cant find any documentation on `Dll_GetProxyDllInfo` How do you know the above is the case?